src/wayland/virtual_keyboard/virtual_keyboard_handle.rs

```diff
@@ -0,0 +1,205 @@
+    pub(super) handle: VirtualKeyboardHandle,
+    pub(crate) seat: Seat<D>,
+}
+
+impl<D: SeatHandler> Debug for VirtualKeyboardUserData<D>
+where
+    <D as SeatHandler>::KeyboardFocus: Debug,
+    <D as SeatHandler>::PointerFocus: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VirtualKeyboardUserData")
+            .field("handle", &self.handle)
+            .field("seat", &self.seat.arc)
+            .finish()
+    }
+}
+
+impl<D> Dispatch<ZwpVirtualKeyboardV1, VirtualKeyboardUserData<D>, D> for VirtualKeyboardManagerState
+where
+    D: Dispatch<ZwpVirtualKeyboardV1, VirtualKeyboardUserData<D>>,
+    D: SeatHandler + 'static,
+    <D as SeatHandler>::KeyboardFocus: WaylandFocus,
+{
+    fn request(
+        _data: &mut D,
+        _client: &Client,
+        _: &ZwpVirtualKeyboardV1,
+        request: zwp_virtual_keyboard_v1::Request,
+        data: &VirtualKeyboardUserData<D>,
+        _dh: &DisplayHandle,
+        _data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            zwp_virtual_keyboard_v1::Request::Keymap { format, fd, size } => {
+                // This should be wl_keyboard::KeymapFormat::XkbV1,
+                // but the protocol does not state the parameter is an enum.
+                if format == 1 {
+                    let keyboard_handle = data.seat.get_keyboard().unwrap();
+                    let mut internal = keyboard_handle.arc.internal.lock().unwrap();
+                    let old_keymap = internal.keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1);
+                    let context = xkb::Context::new(xkb::CONTEXT_NO_FLAGS);
+                    let new_keymap = unsafe {
+                        xkb::Keymap::new_from_fd(
+                            &context,
+                            fd,
+                            size as usize,
+                            format,
+                            xkb::KEYMAP_COMPILE_NO_FLAGS,
+                        )
+                        .unwrap()
+                        .unwrap()
+                    };
+                    if old_keymap != new_keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1) {
+                        let mut inner = data.handle.inner.lock().unwrap();
+                        let log = crate::slog_or_fallback(None);
+                        if inner.old_keymap.is_none() {
+                            let old_keymap = CString::new(old_keymap)
+                                .expect("Keymap should not contain interior null bytes");
+                            inner.old_keymap = Some(KeymapFile::new(old_keymap, log));
+                        }
+                        internal.keymap = new_keymap;
+                        let known_kbds = &keyboard_handle.arc.known_kbds;
+                        for kbd in &*known_kbds.lock().unwrap() {
+                            kbd.keymap(KeymapFormat::XkbV1, fd, size);
+                        }
```

Comment:
Not saying if this is right or not, but what does happen if someone presses a key on a physical keyboard while the virtual keyboard is open? Won't the map be wrong?

Reply:
yes, it will be the keymap from the virtual keyboard. So yes the map will be "wrong". The alternative would be to keep all keymaps and set the keymaps on all clients before sending the key.... That would be a lot of work.
In sway they only change it when it is different. My take on this, is that in practice virtual keyboards will set the same keymap as the one set on the compositor(aka no change), also input methods will probably set a different one in many cases but that will be intended behavior and thus the physical keyboard will output the "correct" map.
I do set the keymap back to the original one when the virtual one is dropped, which I would guess at least makes it somewhat predictable...
I am open for suggestions :smile:

Reply:
Sounds like a reasonable compromise, I'm fine with that.